File: tools/spellcheck_vocab.js
/* =====================================================
   FUGA MYTHICA — dictionary check of every Cursus word against hunspell dictionaries
   node tools/spellcheck_vocab.js <dictDir> <nspellModuleDir> [--out report.json]
   dictDir holds en_GB / fr / es_ES / it_IT / de_DE_frami / la  .dic+.aff
   Strips articles (der/die/das, le/la/les/l', el/la/los/las, il/lo/la/i/gli/le/l') and checks each remaining token.
   Multi-word phrases: every token must pass. Reports words with any unknown token.
   ===================================================== */
const fs = require("fs"), vm = require("vm"), path = require("path");
const [dictDir, nspellDir] = process.argv.slice(2);
const nspell = require(path.join(nspellDir, "nspell"));
const ROOT = path.join(__dirname, "..");
const html = fs.readFileSync(path.join(ROOT, "index.html"), "utf8");
const files = [...html.matchAll(/<script src="([^"]+\.js)"/g)].map(m => m[1]).filter(f => /^(data|vocab)/.test(f));
const ctx = {}; vm.createContext(ctx);
for (const f of files) vm.runInContext(fs.readFileSync(path.join(ROOT, f), "utf8"), ctx, { filename: f });
const V = vm.runInContext("VOCAB", ctx); V.la = vm.runInContext("buildLatinPack()", ctx);
const DICT = { en: "en_GB", fr: "fr", es: "es_ES", it: "it_IT", de: "de_DE_frami", la: "la" };
const ART = { de: /^(der|die|das|ein|eine)$/, fr: /^(le|la|les|l'|un|une|d'|de|du|des|à|au|aux|se|s'|en|y)$/i, es: /^(el|la|los|las|un|una|de|del|a|al|se|en|por|con)$/i, it: /^(il|lo|la|i|gli|le|l'|un|uno|una|di|del|della|a|al|alla|da|si|in|con|per)$/i, en: /^(the|a|an|to|of|in|on|at|by|for|with)$/i, la: /^$/ };
const report = {};
for (const [lang, name] of Object.entries(DICT)) {
  const aff = fs.readFileSync(path.join(dictDir, name + ".aff")), dic = fs.readFileSync(path.join(dictDir, name + ".dic"));
  const sp = nspell(aff, dic);
  const seen = new Set(); const bad = [];
  for (const w of V[lang].words) {
    if (seen.has(w[0])) continue; seen.add(w[0]);
    const toks = w[0].replace(/[’]/g, "'").split(/[\s\-–—,;:!?()…]+/).flatMap(t => t.includes("'") ? t.split(/(?<=')/) : [t]).map(t => t.trim()).filter(Boolean);
    const unknown = toks.filter(t => !ART[lang].test(t) && !/^\d+$/.test(t) && !/^[A-Z]/.test(t) && !sp.correct(t) && !sp.correct(t.toLowerCase()) && !sp.correct(t.charAt(0).toUpperCase() + t.slice(1)));
    if (unknown.length) bad.push([w[0], unknown.join("|"), w[3] || ""]);
  }
  report[lang] = { checked: seen.size, unknown: bad.length, items: bad };
  console.log(lang, "checked", seen.size, "unknown", bad.length);
}
const out = process.argv.includes("--out") ? process.argv[process.argv.indexOf("--out") + 1] : null;
if (out) fs.writeFileSync(out, JSON.stringify(report, null, 1));

File: tools/spellcheck_vocab.py
"""
FUGA MYTHICA — dictionary check of every Cursus word (spylls = pure-Python hunspell)
  python tools/spellcheck_vocab.py <dictDir> [--out report.json] [--lang en,fr]
dictDir holds en_GB / fr / es_ES / it_IT / de_DE_frami / la  .dic+.aff (LibreOffice / wooorm dictionaries)
Step 1 dumps the packs with node (tools/dump_words.js); step 2 checks every token not an article / number / proper noun.
"""
import sys, json, re, subprocess, os
from spylls.hunspell import Dictionary
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
dictDir = sys.argv[1]
out = sys.argv[sys.argv.index("--out") + 1] if "--out" in sys.argv else None
only = sys.argv[sys.argv.index("--lang") + 1].split(",") if "--lang" in sys.argv else None
words = json.loads(subprocess.check_output(["node", os.path.join(ROOT, "tools", "dump_words.js")], cwd=ROOT).decode("utf-8"))
DICT = {"en": "en_GB", "fr": "fr", "es": "es_ES", "it": "it_IT", "de": "de_DE_frami", "la": "la"}
ART = {"de": r"^(der|die|das|ein|eine|sich|zu)$", "fr": r"^(le|la|les|l'|un|une|d'|de|du|des|à|au|aux|se|s'|en|y|qu')$", "es": r"^(el|la|los|las|un|una|de|del|a|al|se|en|por|con|lo)$",
       "it": r"^(il|lo|la|i|gli|le|l'|un|uno|una|di|del|della|a|al|alla|da|si|in|con|per|dei|delle|degli)$", "en": r"^(the|a|an|to|of|in|on|at|by|for|with)$", "la": r"^$"}
report = {}
for lang, name in DICT.items():
    if only and lang not in only: continue
    d = Dictionary.from_files(os.path.join(dictDir, name))
    art = re.compile(ART[lang], re.I)
    bad = []
    for w in words[lang]:
        toks = []
        for t in re.split(r"[\s\-–—,;:!?()…]+", w[0].replace("’", "'")):
            toks += [x for x in re.split(r"(?<=')", t) if x]
        unk = [t for t in toks if t and not art.match(t) and not t.isdigit() and not t[0].isupper()
               and not (d.lookup(t) or d.lookup(t.lower()) or d.lookup(t[0].upper() + t[1:]))]
        if unk: bad.append([w[0], "|".join(unk), w[3] or ""])
    report[lang] = {"checked": len(words[lang]), "unknown": len(bad), "items": bad}
    print(lang, "checked", len(words[lang]), "unknown", len(bad), flush=True)
if out:
    json.dump(report, open(out, "w", encoding="utf-8"), ensure_ascii=False, indent=1)
